src/rotation_types.jl

```diff
@@ -0,0 +1,291 @@
+
+trans_dict[RotMatrix] = nothing  # the blessed type, everything goes through RotMatrixs
+
+
+# Quaternions have an extra bool field that type_methods.jl doesn't want to know about
+numel(::Type{RotMatrix}) = 9
+
+strip_eltype{T <: RotMatrix}(::Type{T}) = RotMatrix
+
+
+###################################################
+# Quaternions
+# (Quaternions are defined in Quaternions.jl)
+###################################################
+
+# an extra constructor so we don't need to specify whether its normalized
+@inline call{T}(::Type{Quaternion{T}}, a::Real, b::Real, c::Real, d::Real) = Quaternion(T(a), T(b), T(c), T(d))
+
+# define its interaction with RotMatrixs
+@inline convert(::Type{RotMatrix}, q::Quaternion) = quat_to_rot(q)
+@inline convert{T}(::Type{RotMatrix{T}}, q::Quaternion) = convert(RotMatrix{T}, quat_to_rot(q))
+
+@inline convert(::Type{Quaternion}, R::RotMatrix) = rot_to_quat(R)
+@inline convert{T}(::Type{Quaternion{T}}, R::RotMatrix) = convert(Quaternion{T}, rot_to_quat(R))
+
+# go from a Quaternion to any other representation via the Rotation Matrix
+# @inline convert{T}(::Type{T}, q::Quaternion) = convert(T, quat_to_rot(q))  # go via the rotation representation
+trans_dict[Quaternion] = RotMatrix
+
+# define its interaction with other angle representations
+@inline eltype{T}(::Type{Quaternion{T}}) = T
+@inline eltype{T}(::Quaternion{T}) = T
+
+# Quaternions have an extra bool field that type_methods.jl doesn't want to know about
+numel(::Type{Quaternion}) = 4
+@inline getindex(X::Quaternion, i::Integer) = X.(i)
+
+strip_eltype{T <: Quaternion}(::Type{T}) = Quaternion
+
+
+###################################################
+# Stereographic projection of Quaternions
+###################################################
+
+"""
+
+A 3 element parametrization of a unit quaternion Q formed by the intersection of a line from [-1,0,0,0] to Q, with a plane containing the origin and with normal direction [1,0,0,0]. This
+is a compact representation of rotations where the derivitives of the rotation matrix's elements wrt the SpQuat parameters are rational functions (good for optimization)
+
+See:
+
+    [1] "A Recipe on the Parameterization of Rotation Matrices for Non-Linear Optimization using Quaternions",
+    Terzakis, Culverhouse, Bugmann, Sharma, Sutton,
+    MIDAS technical report, 2011
+    http://www.tech.plymouth.ac.uk/sme/springerusv/2011/publications_files/Terzakis%20et%20al%202012,%20A%20Recipe%20on%20the%20Parameterization%20of%20Rotation%20Matrices...MIDAS.SME.2012.TR.004.pdf
+
+    Note 1: the singularity (origin) has been moved from [0,0,0,-1] in Ref[1] to [-1,0,0,0], so the 0 rotation quaternion [1,0,0,0] maps to [0,0,0] as opposed of to [1,0,0].
+    Note 2: the quaternion rotation ambiguity q = -q means there will be a rotation with ||SpQuat|| <= 1 and an equivilent rotation with ||SpQuat|| > 1.  This package
+            uses the solution with ||SpQuat|| <= 1
```

Comment:
<3

On 10 May 2016 at 15:15, Chris Foster <notifications@github.com> wrote:

> In src/rotation_types.jl
> <https://github.com/FugroRoames/Rotations.jl/pull/2#discussion_r62613415>:
>
> > +
> > +"""
> > +
> > +A 3 element parametrization of a unit quaternion Q formed by the intersection of a line from [-1,0,0,0] to Q, with a plane containing the origin and with normal direction [1,0,0,0]. This
> > +is a compact representation of rotations where the derivitives of the rotation matrix's elements wrt the SpQuat parameters are rational functions (good for optimization)
> > +
> > +See:
> > +
> > +    [1] "A Recipe on the Parameterization of Rotation Matrices for Non-Linear Optimization using Quaternions",
> > +    Terzakis, Culverhouse, Bugmann, Sharma, Sutton,
> > +    MIDAS technical report, 2011
> > +    http://www.tech.plymouth.ac.uk/sme/springerusv/2011/publications_files/Terzakis%20et%20al%202012,%20A%20Recipe%20on%20the%20Parameterization%20of%20Rotation%20Matrices...MIDAS.SME.2012.TR.004.pdf
> > +
> > +    Note 1: the singularity (origin) has been moved from [0,0,0,-1] in Ref[1] to [-1,0,0,0], so the 0 rotation quaternion [1,0,0,0] maps to [0,0,0] as opposed of to [1,0,0].
> > +    Note 2: the quaternion rotation ambiguity q = -q means there will be a rotation with ||SpQuat|| <= 1 and an equivilent rotation with ||SpQuat|| > 1.  This package
> > +            uses the solution with ||SpQuat|| <= 1
>
> Nice.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/FugroRoames/Rotations.jl/pull/2/files/fb3a7882257b93afed97cfaa9bf56669bbf410dc#r62613415>
>